src/wasm-binary.h

```diff
@@ -792,33 +866,651 @@ class WasmBinaryWriter : public WasmVisitor<void> {
+
+  void readExpression(Expression*& curr) {
+    if (debug) std::cerr << "zz recurse into " << ++depth << " at " << pos << std::endl;
+    uint8_t code = getInt8();
+    if (debug) std::cerr << "readExpression seeing " << (int)code << std::endl;
+    switch (code) {
+      case BinaryConsts::Block:        visitBlock((curr = allocator.alloc<Block>())->cast<Block>()); break;
+      case BinaryConsts::If:
+      case BinaryConsts::IfElse:       visitIf((curr = allocator.alloc<If>())->cast<If>(), code);  break;// code distinguishes if from if_else
+      case BinaryConsts::Loop:         visitLoop((curr = allocator.alloc<Loop>())->cast<Loop>()); break;
+      case BinaryConsts::Br:
+      case BinaryConsts::BrIf:         visitBreak((curr = allocator.alloc<Break>())->cast<Break>(), code); break; // code distinguishes br from br_if
+      case BinaryConsts::TableSwitch:  visitSwitch((curr = allocator.alloc<Switch>())->cast<Switch>()); break;
+      case BinaryConsts::CallFunction: {
+        // might be an import or not. we have to check here.
+        Name target = mappedFunctions[getLEB128()];
+        assert(target.is());
+        if (debug) std::cerr << "call(import?) target: " << target << std::endl;
+        if (wasm.importsMap.find(target) == wasm.importsMap.end()) {
+          assert(wasm.functionsMap.find(target) != wasm.functionsMap.end());
+          visitCall((curr = allocator.alloc<Call>())->cast<Call>(), target);
+        } else {
+          visitCallImport((curr = allocator.alloc<CallImport>())->cast<CallImport>(), target);
+        }
+        break;
+      }
+      case BinaryConsts::CallIndirect: visitCallIndirect((curr = allocator.alloc<CallIndirect>())->cast<CallIndirect>()); break;
+      case BinaryConsts::GetLocal:     visitGetLocal((curr = allocator.alloc<GetLocal>())->cast<GetLocal>()); break;
+      case BinaryConsts::SetLocal:     visitSetLocal((curr = allocator.alloc<SetLocal>())->cast<SetLocal>()); break;
+      case BinaryConsts::Select:       visitSelect((curr = allocator.alloc<Select>())->cast<Select>()); break;
+      case BinaryConsts::Nop:          visitNop((curr = allocator.alloc<Nop>())->cast<Nop>()); break;
+      case BinaryConsts::Unreachable:  visitUnreachable((curr = allocator.alloc<Unreachable>())->cast<Unreachable>()); break;
+      default: {
+        // otherwise, the code is a subcode TODO: optimize
+        if (maybeVisit<Binary>(curr, code)) break;
+        if (maybeVisit<Unary>(curr, code)) break;
+        if (maybeVisit<Const>(curr, code)) break;
+        if (maybeVisit<Load>(curr, code)) break;
+        if (maybeVisit<Store>(curr, code)) break;
+        if (maybeVisit<Host>(curr, code)) break;
+        abort();
+      }
+    }
+    if (debug) std::cerr << "zz recurse from " << depth-- << " at " << pos << std::endl;
+  }
+
+  template<typename T>
+  bool maybeVisit(Expression*& curr, uint8_t code) {
+    T temp;
+    if (maybeVisitImpl(&temp, code)) {
+      auto actual = allocator.alloc<T>();
```

Comment:
hmm, this is a super-common pattern across binaryen - calls to `allocator.alloc<T>`, returning a pointer. I don't feel `auto*` adds anything there?